### PR TITLE
Merge stack before printing

### PR DESCRIPTION
**Pull request by zhiburt**
Could you @fdncred try it?

close?: #9264

### CODE REVIEW

**Comment by fdncred** [comment]
@zhiburt Doesn't seem to work. Here's my test.

1. save preview-theme.nu from https://github.com/nushell/nushell/issues/9264#issuecomment-1561357591
2. source preview-theme.nu
3. cd nu_scripts
4. preview <tab> and i randomly picked apprentice
5. which then puts this on the command line `use themes/themes/apprentice.nu; $env.config.color_config = (apprentice); preview_theme | table -e` and hit enter
6. I get this
image
Which you can tell it's not working because the header and row index (which are the same color) `{fg: #87af87, attr: b}` are not being drawn in that color. the green header and cyan index are from my current theme, before i `use`'d the one above.

**Comment by fdncred** [comment]
This is so strange. If I do the commands on separate lines, it seems to work.
image

**Comment by fdncred** [comment]
It seems this may have something to do with how nushell is handling one-liners with multiple commands separated with semicolons? like all the stuff that is changing in this pr may not be happening on this type of one-liner?

**Comment by zhiburt** [comment]
I think must be addressed.
I got lazy a bit here and just followed @kubouch suggestion for this case.

Let me know if it works.

**Comment by fdncred** [comment]
wow, this is working great! still playing but just wow!!!

**Comment by fdncred** [comment]
@zhiburt I'm ready to merge this but can we have more descriptive fn names for `merge_env2` and `merge_env3`?

File: crates/nu-protocol/src/engine/engine_state.rs [hunk 1]
```diff
@@ -310,72 +306,25 @@ impl EngineState {
...
-                    }
-                } else {
-                    // Pushing a new overlay
-                    self.env_vars.insert(overlay_name, env);
-                }
-            }
-        }
-
+    /// Set a CWD.
+    pub fn set_current_workink_dir(&mut self, cwd: impl AsRef<Path>) -> Result<(), ShellError> {
```
Comment (fdncred) [inline crates/nu-protocol/src/engine/engine_state.rs:313]:
```suggestion
    pub fn set_current_working_dir(&mut self, cwd: impl AsRef<Path>) -> Result<(), ShellError> {
```

**Comment by fdncred** [comment]
Thanks @zhiburt!

**Comment by kubouch** [comment]
I didn't have time to look into it yet but we shouldn't clone the EngineState before every print, that will be super slow.

**Comment by zhiburt** [comment]
Precisely `clone` will make it slower.
The same for `table`.

I am not sure to which degree though; you could test it.

Ideally the `env` shall be just merged somewhere in `nu-engine/eval`; but I haven't cracked this one.

**Pull request merged**